src/index.py

```diff
@@ -79,3 +96,260 @@ def contents_from_page(page: str) -> str:
+    """
+    match = _ITEM_PATTERN.match(line)
+
+    if match is None:
+        raise InputError(
+            f"An item in the contents of the index file at {DOCUMENTATION_INDEX_FILENAME} is "
+            f"invalid, {line=!r}, expecting regex: {_ITEM}"
+        )
+
+    whitespace_count = len(match.group(1))
+
+    if whitespace_count != 0 and rank == 0:
+        raise InputError(
+            f"An item in the contents of the index file at {DOCUMENTATION_INDEX_FILENAME} is "
+            f"invalid, {line=!r}, expecting the first line not to have any leading whitespace"
+        )
+
+    reference_title = match.group(7)
+    reference_value = match.group(8)
+
+    return _ParsedListItem(
+        whitespace_count=whitespace_count,
+        reference_title=reference_title,
+        reference_value=reference_value,
+        rank=rank,
+    )
+
+
+def _remove_contents(lines: typing.Iterable[str]) -> typing.Iterator[str]:
+    """Remove the contents section lines from the index contents.
+
+    Args:
+        lines: The lines of the index file.
+
+    Yields:
+        All lines except the lines of the contents section.
+    """
+    contents_encountered = False
+    drop_lines = False
+    for line in lines:
+        if not contents_encountered and line.lower() == CONTENTS_HEADER:
+            contents_encountered = True
+            drop_lines = True
+        elif line.startswith(CONTENTS_END_LINE_PREFIX):
+            drop_lines = False
+
+        if not drop_lines:
+            yield line
+
+
+def get_content_for_server(index_file: IndexFile) -> str:
+    """Get the contents from the index file that should be passed to the server.
+
+    Args:
+        index_file: Information about the local index file.
+
+    Returns:
+        The contents of the index file that should be stored on the server.
+    """
+    if index_file.content is None:
+        return ""
+
+    return "\n".join(_remove_contents(index_file.content.splitlines()))
+
+
+def _get_contents_parsed_items(index_file: IndexFile) -> typing.Iterator[_ParsedListItem]:
+    """Get the items from the contents list of the index file.
+
+    Args:
+        index_file: The index file to read the contents from.
+
+    Yields:
+        All the items on the contents list in the index file.
+    """
+    if index_file.content is None:
+        return
+
+    # Get the lines of the contents section
+    lines = index_file.content.splitlines()
+    # Advance past the contents heading
+    lines_from_contents = itertools.dropwhile(lambda line: line.lower() != CONTENTS_HEADER, lines)
+    next(lines_from_contents, None)
+    # Stop taking on the next heading
+    contents_lines = itertools.takewhile(
+        lambda line: not line.startswith(CONTENTS_END_LINE_PREFIX), lines_from_contents
+    )
+    yield from (
+        _parse_item_from_line(line=line, rank=rank)
+        for line, rank in zip(filter(None, contents_lines), itertools.count())
+    )
+
+
+def _check_contents_item(
```

Comment:
There's no check on the coherence of numbering or lettering, so something like:
```
- [title 1](value 1)
  3. [title 1.1](value 1.1)
  2. [title 1.2](value 1.2)
```
Or its equivalent with letters, will just keep the line ordering ignoring the letter or number coherence

Reply:
yes, similar as other comments, that should be left to markdown linters

Reply:
As discussed on the other comment, this can be addressed using a linter check in a follow up PR.

In general, GitHub allows e.g., 

3. ...
2. ...

and will just correct on rendering